<div style="border:solid green 2px; padding: 20px">
<b>Андрей, привет!</b>

Меня зовут Евгений Дащенко, и я буду проверять твой проект. Предлагаю общаться на «ты» 😊 Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b> Так выделены небольшие замечания. Одна-два таких замечания в проекте допустимы, но если их будет больше - тебе будет необходимо внести исправления. Это как тестовое задание при приеме на работу: очень много мелких ошибок могут стать причиной отказа кандидату. 

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
    

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-к-базе" data-toc-modified-id="Подключение-к-базе-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе</a></span></li><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#Аналитика-данных" data-toc-modified-id="Аналитика-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Аналитика данных</a></span><ul class="toc-item"><li><span><a href="#Cколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Cколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Cколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Какая-средняя-оценка-и-количество-обзоров-у-каждой-книги" data-toc-modified-id="Какая-средняя-оценка-и-количество-обзоров-у-каждой-книги-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Какая средняя оценка и количество обзоров у каждой книги</a></span></li><li><span><a href="#Какое-издательство,-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Какое-издательство,-выпустило-наибольшее-число-книг-толще-50-страниц-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Какое издательство, выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Какое-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок?" data-toc-modified-id="Какое-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Какое среднее количество обзоров от пользователей, которые поставили больше 50 оценок?</a></span></li><li><span><a href="#Какой-автор-с-самой-высокой-средней-оценкой-книг?" data-toc-modified-id="Какой-автор-с-самой-высокой-средней-оценкой-книг?-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Какой автор с самой высокой средней оценкой книг?</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Книжное дело

Компанией было приобретен крупный сервис для чтения книг по подписке. Необходимо проанализировать имеющуюся информацию и определить направления развития

# Цель:
Пронализировать базу данных, в которой находятся данные о книгах с целью формулирования ценностного предложения для нового продукта.

# Исходные данные:
Информация о книгах, издательствах, авторах, а также пользовательские обзоры книг
База данных, содержащая таблицы: books, authors, publishers, ratings и reviews

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Здорово, когда есть вступление в сам проект, каждый сможет быстрее разобраться в концепции / деталях работы. 

</div>

## Подключение к базе

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config = {
    'user': 'praktikum_student',
    'pwd': 'Sdf4$2;d-d30pp',
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432, 
    'db': 'data-analyst-final-project-db'} 
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Коннекшн готов, теперь можем делать запросы 😊
</div>

In [2]:
sql='''
SELECT *
FROM books
LIMIT 5
'''

In [3]:
pd.io.sql.read_sql(sql, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Сделаем запрос в виде функции для удобства

In [4]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Отличная функция для выполнения запросов 👍
</div>

In [5]:
select(sql)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Работает, отлично

## Изучение данных

In [6]:
sql='''
SELECT *
FROM books b
LIMIT 5
'''

In [7]:
select(sql)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [8]:
sql='''
SELECT count(b.book_id) AS cnt
FROM books b
'''

In [9]:
select(sql)

cnt
0  1000

Таблица book содержит 1000 записей

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [10]:
sql='''
SELECT *
FROM authors a
LIMIT 5
'''

In [11]:
select(sql)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [12]:
sql='''
SELECT count(a.author_id) AS cnt
FROM authors a
'''

In [13]:
select(sql)

cnt
0  636

Таблица autors содержит 636 записей

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [14]:
sql='''
SELECT *
FROM publishers p
LIMIT 5
'''

In [15]:
select(sql)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [16]:
sql='''
SELECT count(p.publisher_id) AS cnt
FROM publishers p
'''

In [17]:
select(sql)

cnt
0  340

Таблица publisher содержит 340 записей

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [18]:
sql='''
SELECT *
FROM ratings r
LIMIT 5
'''

In [19]:
select(sql)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [20]:
sql='''
SELECT count(r.rating_id) AS cnt
FROM ratings r
'''

In [21]:
select(sql)

cnt
0  6456

Таблица  rating содержит 6456 записей

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [22]:
sql='''
SELECT *
FROM reviews rev
LIMIT 5
'''

In [23]:
select(sql)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [24]:
sql='''
SELECT count(rev.review_id) AS cnt
FROM reviews rev
'''

In [25]:
select(sql)

cnt
0  2793

Таблица reviews содержит 2793 записей.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Что ж, данные посмотрели теперь мы можем приступить к выполнению наших заданий.
    
Кстати получение данных можно делать следующим образом: 
    
`pd.read_sql(query, con)`
    
</div>

## Аналитика данных

### Cколько книг вышло после 1 января 2000 года

In [26]:
sql='''
SELECT count(b.book_id) AS cnt
FROM books b
WHERE publication_date > '2000-01-01'
'''

In [27]:
select(sql)

cnt
0  819

После 1 января вышло 819 книг

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b>

Обрати внимание, что язык `SQL` работает в строгом формате при выполнении логических фильтраций `>` и прочих знаков.
    
В данном случае, чтобы не терять лишние данные нам стоит указать `>=`, хоть и формулировка немного неточная в нашем задании.

    
Оставил пример ниже.
    
</div>

### Какая средняя оценка и количество обзоров у каждой книги

In [28]:
sql='''
WITH average_rating AS
  (SELECT b.book_id,
          title,
          avg(r.rating) AS average_rating
   FROM books b
   LEFT JOIN ratings r ON r.book_id = b.book_id
   GROUP BY b.book_id),
     count_reviews AS
  (SELECT b.book_id,
          title,
          count(rev.review_id) AS count_reviews
   FROM books b
   LEFT JOIN reviews rev ON rev.book_id = b.book_id
   GROUP BY b.book_id)
SELECT a.book_id,
       a.title, 
       c.count_reviews,
       a.average_rating
FROM average_rating a
JOIN count_reviews c ON a.book_id = c.book_id
ORDER BY a.average_rating DESC
LIMIT 5
'''

In [29]:
select(sql)

book_id                                      title  count_reviews  \
0       55  A Woman of Substance (Emma Harte Saga #1)              2   
1       57            Act of Treason (Mitch Rapp  #9)              2   
2       17              A Dirty Job (Grim Reaper  #1)              4   
3       20      A Fistful of Charms (The Hollows  #4)              2   
4       62                              Alas  Babylon              2   

   average_rating  
0             5.0  
1             5.0  
2             5.0  
3             5.0  
4             5.0

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Супер, все посчитано корректно 👍
    
</div>

In [30]:
sql='''
SELECT avg(r.rating)
FROM ratings r
WHERE r.book_id = 55
'''

In [31]:
select(sql)

avg
0  5.0

In [32]:
sql='''
SELECT count(rev.review_id)
FROM reviews rev
WHERE rev.book_id = 57
'''

In [33]:
select(sql)

count
0      2

Наивысшая оценка - 5, наименьшая - 2. Для каждой из книг мы определили количество обзоров

### Какое издательство, выпустило наибольшее число книг толще 50 страниц

In [34]:
sql='''
SELECT p.publisher_id,
       p.publisher,
       count(book_id) AS cnt
FROM books b
JOIN publishers p ON p.publisher_id = b.publisher_id
WHERE num_pages > 50
GROUP BY p.publisher_id,
         p.publisher
HAVING count(book_id) =
  (SELECT max(cnt)
   FROM
     (SELECT count(book_id) AS cnt
      FROM books b
      WHERE num_pages > 50
      GROUP BY b.publisher_id) AS t)
'''

In [35]:
select(sql)

publisher_id      publisher  cnt
0           212  Penguin Books   42

Издательство Penguin Books выпустило больше всего книг (42) толщиной больше 50 страниц

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Издательство определено верно 👍
    
Здорово, что ключевые слова языка `SQL` отображаешь в верхнем регистре, так быстрее понять структуру запроса.
    
---
    
Оставил пример ниже, как это можно было посчитать еще.
    
</div>

In [36]:
pd.read_sql("""SELECT p.publisher, COUNT(b.book_id) as books_count FROM books b 
LEFT JOIN publishers p ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY books_count DESC
LIMIT 1
""", con = engine)

publisher  books_count
0  Penguin Books           42

### Какое среднее количество обзоров от пользователей, которые поставили больше 50 оценок?

In [37]:
sql='''
WITH users_more_50 AS
  (SELECT r.username
   FROM ratings r
   GROUP BY r.username
   HAVING (count(r.rating_id) > 50))
SELECT avg(t.cnt) AS avg_review
FROM
  (SELECT count(rev.review_id) AS cnt
   FROM reviews rev
   WHERE rev.username in
       (SELECT *
        FROM users_more_50)
   GROUP BY rev.username) AS t
'''

In [38]:
select(sql)

avg_review
0   24.333333

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Все посчитано верно 👍
    
</div>

В среднем, пользователь, поставивший больше 50-ти оценок составляет 24 обзора на книги

### Какой автор с самой высокой средней оценкой книг?

In [52]:
sql='''
SELECT a.author, r.avg_rating
FROM books b
LEFT JOIN 
(select book_id, avg(rating) avg_rating, count(rating) count_rating
from ratings
group by book_id) r 
on r.book_id=b.book_id
LEFT JOIN authors a on a.author_id=b.author_id
--GROUP BY a.author
WHERE r.count_rating>=50
ORDER BY r.avg_rating desc
LIMIT 1
'''

In [53]:
select(sql)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.414634

Автор с самой высокой средней оценкой книг это 	J.K. Rowling

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>

Наши расчеты получились с дублями, т.к. мы к одной таблице присоединяем остальные 2 и соответственно количество данных увеличивается где нет совпадений через `LEFt JOIN`. Кстати у тебя посчитано количество через `count`, а нам нужен средний рейтинг автора.
    
Проблема также в том, что нам нужно сделать все сгруппированные расчеты заранее по таблицам ревью и рейтингов и присоединить их к таблице `books` также с помощью `LEFt JOIN` и выбрать просто посчитанные значения и книги, тогда дублей у нас не должно возникнуть и будет посчитано все корректно.
    
Давай поправим этот момент.
    
</div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Комментарий студента:</b> Жень, привет!Да, согласен, поправил, теперь вроде правильно) Не думал, что тут тоже надо писать выводы, просто в задании этого нет, но сейчас напишу)</div>

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>

Финальный вывод и есть главный результат твоей работы. Стоит писать его подробно по результатам проведенной работы. В нем можно приводить полученные в ходе работы значения. Также можно расписать все, что было сделано в работе.
    
Самое главное — это вывод должен содержать в себе ответ на поставленную задачу.

</div>

## Выводы

В ходе анализа базы данных, содержащей информацию о книгах (в том числе об авторах, издательствах, оценках и обзорах), были определены следующие факты:


- 819 книг из 1000 были изданы в 21 веке (после 1 января 2000)
- максимальная оценка книги является 5, а минимальная (из имеющихся) 2
- 42 книги выпустило издательство Penguin Books, что является наибольшим значением среди всех издательств (брошюры не учитывались)
- автором с самыми высокими оценками является J.K. Rowling
- пользователи, оставляющие более 50 оценок к книгам, в среднем делают по 24 обзора